### PR TITLE
Ceph Bucket management

### PR DESCRIPTION
**Pull request by Click2cloud-Pallas**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind new feature
> /kind bug fix
> /kind cleanup
> /kind revert change
> /kind design
> /kind documentation
> /kind enhancement

**What this PR does / why we need it**:
This PR adds support for Ceph bucket management
**Which issue(s) this PR fixes**:

Fixes #

**Test Report Added?**:
> kind TESTED

**Test Report**:
CreateBackend
createbackend
BackendCreateSucessful

CreateBucket

bucketsoda
ObjectUpload
objectuploadsoda
AzureToCephMigration
azuretocephmig
BucketCreatedCheckCeph
bucketpreviewceph
MigratedAndUploadedObjectCheckCeph
objectpreviewceph

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by PravinRanjan10** [comment]
LGTM

File: s3/pkg/datastore/ceph/ceph.go [hunk 1]
```diff
@@ -34,22 +35,38 @@ import (
...
 }
 
 func (ad *CephAdapter) BucketDelete(ctx context.Context, in *pb.Bucket) error {
+	bucket := ad.session.NewBucket()
+	err := bucket.Remove(in.Name)
+	if err != nil {
+		log.Error("the Delete bucket failed in ceph service with err:%s", err.Error())
+		return err
+	}
+	log.Debug("The bucket:%s deleted successful in ceph", in.Name)
```
Comment (anvithks) [inline s3/pkg/datastore/ceph/ceph.go:52]: nit: successful -> successfully

File: s3/pkg/datastore/ceph/ceph.go [hunk 2]
```diff
@@ -34,22 +35,38 @@ import (
...
 }
 
 func (ad *CephAdapter) BucketCreate(ctx context.Context, input *pb.Bucket) error {
+	bucket := ad.session.NewBucket()
+	err := bucket.Create(input.Name, models.Private)
+	if err != nil {
+		log.Error("the create bucket failed in ceph service with err:%s", err.Error())
+		return err
+	}
+	log.Debug("The bucket:%s creation successful in ceph", input.Name)
```
Comment (anvithks) [inline s3/pkg/datastore/ceph/ceph.go:63]: nit: following the same tense as above. -> The bucket :%s **created** **successfully**...

**Review by anvithks — APPROVED** [review]
Overall LGTM. Some minor nits.
Please check.

File: s3/pkg/datastore/ceph/ceph.go (hunk 1 above)
Comment (Click2cloud-Pallas) [reply s3/pkg/datastore/ceph/ceph.go]: changes done

File: s3/pkg/datastore/ceph/ceph.go (hunk 2 above)
Comment (Click2cloud-Pallas) [reply s3/pkg/datastore/ceph/ceph.go]: @anvithks done with the changes

File: s3/pkg/datastore/ceph/ceph.go [hunk 3]
```diff
@@ -34,22 +35,38 @@ import (
...
 type CephAdapter struct {
 	backend *backendpb.BackendDetail
 	session *s3client.Client
 }
 
 func (ad *CephAdapter) BucketDelete(ctx context.Context, in *pb.Bucket) error {
+	bucket := ad.session.NewBucket()
+	err := bucket.Remove(in.Name)
+	if err != nil {
+		log.Error("the Delete bucket failed in ceph service with err:%s", err.Error())
```
Comment (kumarashit) [inline s3/pkg/datastore/ceph/ceph.go:49]: nit: 
```suggestion
		log.Error("the delete bucket failed in ceph service with err: %s", err.Error())
```

File: s3/pkg/datastore/ceph/ceph.go [hunk 4]
```diff
@@ -319,7 +336,26 @@ func (ad *CephAdapter) ListParts(ctx context.Context, multipartUpload *pb.ListPa
 }
 
 func (ad *CephAdapter) BackendCheck(ctx context.Context, backendDetail *pb.BackendDetailS3) error {
-    return ErrNotImplemented
+	randId := uuid.NewV4().String()
+	input := &pb.Bucket{
+		Name: sampleBucket + randId,
+	}
+
+	err := ad.BucketCreate(ctx, input)
```
Comment (kumarashit) [inline s3/pkg/datastore/ceph/ceph.go:344]: Is BucketCreate synchronous call? 
If it's async, immediate bucketDelete may through error

Comment (Click2cloud-Pallas) [reply s3/pkg/datastore/ceph/ceph.go:344]: ok @kumarashit will check and update the code

File: s3/pkg/datastore/ceph/ceph.go (hunk 3 above)
Comment (Click2cloud-Pallas) [reply s3/pkg/datastore/ceph/ceph.go]: Done changes

File: s3/pkg/datastore/ceph/ceph.go (hunk 4 above)
Comment (Click2cloud-Pallas) [reply s3/pkg/datastore/ceph/ceph.go:344]: @kumarashit I have updated the code and handled the issue by calling Head bucket

File: s3/pkg/datastore/ceph/ceph.go [hunk 5]
```diff
@@ -319,7 +336,34 @@ func (ad *CephAdapter) ListParts(ctx context.Context, multipartUpload *pb.ListPa
...
+	if err != nil {
+		log.Error("failed to create sample bucket :", err)
+		return err
+	}
+	bucket := ad.session.NewBucket()
+	bucketResp, err1 := bucket.Get(input.Name, "", "", "", 1000)
```
Comment (kumarashit) [inline s3/pkg/datastore/ceph/ceph.go:350]: There is no retry?

File: s3/pkg/datastore/ceph/ceph.go [hunk 6]
```diff
@@ -319,7 +336,42 @@ func (ad *CephAdapter) ListParts(ctx context.Context, multipartUpload *pb.ListPa
 }
 
 func (ad *CephAdapter) BackendCheck(ctx context.Context, backendDetail *pb.BackendDetailS3) error {
-    return ErrNotImplemented
+	randId := uuid.NewV4().String()
+	input := &pb.Bucket{
+		Name: sampleBucket + randId,
+	}
+	for i := 1; i <= 3; i++ {
```
Comment (kumarashit) [inline s3/pkg/datastore/ceph/ceph.go:343]: What is the need to put retry in bucket create? Any scenario you see?

Comment (Click2cloud-Pallas) [reply s3/pkg/datastore/ceph/ceph.go:343]: yes if there is some issue due to network it will get resolved in 2nd  or maybe third call

**Review by kumarashit — APPROVED** [review]
LGTM

**Pull request merged**